### ISSUE TITLE
[BUG] `Resolve error: unable to load resolver "node"` under npm's isolated install layout (`--install-strategy=linked`)

### ISSUE THREAD

**Issue by manzoorwanijk** [benmosher/eslint-plugin-import#3268]
Installing with npm's isolated layout breaks the plugin's **default** resolver. No custom resolver is configured, and every import — including core modules — is reported unresolved.

This is the "npm isolated mode" case anticipated in [#2283](https://github.com/import-js/eslint-plugin-import/pull/2283) (details at the bottom).

## Minimal reproducible example

Four files, stock config:

`.npmrc`

```ini
install-strategy=linked
```

`package.json`

```json
{
  "type": "module",
  "scripts": {
    "lint": "eslint src"
  },
  "devDependencies": {
    "eslint": "^9.39.5",
    "eslint-plugin-import": "^2.32.0"
  }
}
```

`eslint.config.mjs`

```js
import importPlugin from "eslint-plugin-import";

export default [
  {
    files: ["src/**/*.js"],
    languageOptions: { ecmaVersion: 2022, sourceType: "module" },
    plugins: { import: importPlugin },
    rules: { "import/no-unresolved": "error" },
  },
];
```

`src/index.js`

```js
import path from "path";

export default path;
```

## Expected

No errors. `path` is a core module and the config is stock.

## Actual

```
src/index.js
  1:1   error  Resolve error: unable to load resolver "node"  import/no-unresolved
  1:18  error  Unable to resolve path to module 'path'        import/no-unresolved

✖ 2 problems (2 errors, 0 warnings)
```

The second error is a cascade of the first, not a separate issue: `requireResolver` throws, `resolve()` catches it and returns `undefined`, and `no-unresolved` treats `undefined` as "not found". Because core-module classification lives inside the resolver that never loaded, **every** import is reported — builtins included. `Resolve error` is deduped once per context via `erroredContexts`, while `Unable to resolve` fires once per import, so a real project gets one cause line and an error on every import in every file.

## Debug output

`DEBUG=eslint-plugin-import:*` produces **no output** for this failure — `utils/resolve.js` contains no `debug` calls, so the resolver-loading path is entirely unlogged. Widening the namespace yields only:

```
$ DEBUG='eslint-plugin-import:*,eslint-module-utils:*' npx eslint src
2026-07-16T10:37:59.295Z eslint-module-utils:ModuleCache cache miss for .../src44136fa...path
```

## Cause

`eslint-import-resolver-node` is a dependency of **`eslint-plugin-import`**, but the code that loads it — `tryRequire` in **`eslint-module-utils/resolve.js`** — lives in a different package that does not declare it. `tryRequire` gets two chances, and under an isolated layout both miss:

| require anchored at                             | finds `eslint-import-resolver-node`?       |
| ----------------------------------------------- | ------------------------------------------ |
| the linted source file (`src/index.js`)         | no — the app doesn't depend on it          |
| `eslint-module-utils/resolve.js` (the fallback) | no — utils doesn't declare it              |
| `eslint-plugin-import/lib/index.js`             | **yes** — it's the plugin's own dependency |

Verified directly against the repro's tree:

```
FAIL sourceFile (src/index.js)  [tryRequire arg 2] -> MODULE_NOT_FOUND
FAIL resolve.js itself          [tryRequire fallback] -> MODULE_NOT_FOUND
OK   plugin lib/index.js -> .../eslint-import-resolver-node/index.js
```

A hoisted layout flattens every package into one top-level `node_modules`, so any package can require any other and the missing declaration is invisible. `--install-strategy=linked` gives each package only its declared dependencies, which exposes it.

## Why this has not been reported via pnpm or yarn

They patch it at install time. Yarn maintains a compatibility database of packages with undeclared dependencies ([`@yarnpkg/extensions`](https://github.com/yarnpkg/berry/blob/0a230c14e71247576f6b51fa811ae08edb6608aa/packages/yarnpkg-extensions/sources/index.ts#L745)); pnpm bundles the same list. It contains this entry, which cites this very PR:

```js
// https://github.com/import-js/eslint-plugin-import/pull/2283
[`eslint-module-utils@*`, {
  peerDependenciesMeta: {
    "eslint-import-resolver-node": optionalPeerDep,
    "eslint-import-resolver-typescript": optionalPeerDep,
    "eslint-import-resolver-webpack": optionalPeerDep,
    "@typescript-eslint/parser": optionalPeerDep
  }
}],
```

Those peer deps do not exist in the published `eslint-module-utils@2.14.0` package.json (`peerDependencies` is `undefined`); pnpm injects them during resolution, so the resolver is symlinked into `eslint-module-utils`'s own `node_modules`. This is not about hoisting — with `hoistPattern: []` (no `.pnpm/node_modules` at all) pnpm still passes. **pnpm and yarn cannot reproduce this**; npm has no such database, so it surfaces only there.

## Relationship to #2283

[#2283](https://github.com/import-js/eslint-plugin-import/pull/2283) (Oct 2021, still open) proposed declaring these as optional peer dependencies and was rejected, with the note:

> For one, anything npm doesn't do isn't something I think needs to be catered to - npm is the standard. That said, npm will be adding a pnpm-like "isolated mode" soon, at which point I'd be obliged to address it somehow.

npm's isolated mode now exists — `--install-strategy=linked` is it, and this report is that case.

Two points distinguish this from the argument in that thread:

1. **This is not a consumer-config problem.** The rejection reasoning was that `eslint-module-utils` dynamically requires resolvers based on config, so consumers must install what their config references. Here nothing references a resolver: `node` is the default, and the plugin already ships `eslint-import-resolver-node` as its own dependency.
2. **A fix need not involve peer dependencies at all,** avoiding the objection that peer deps aren't for declaring "allowed dependencies to access". Since `eslint-plugin-import` is the package that actually depends on the resolver, it can pass its own `require` (or a resolved path) down to `tryRequire`, so the lookup is anchored at the package that owns the dependency.

Happy to open a PR along the lines of (2) if that direction is agreeable.

## Environment

|                             |              |
| --------------------------- | ------------ |
| eslint-plugin-import        | 2.32.0       |
| eslint-module-utils         | 2.14.0       |
| eslint-import-resolver-node | 0.3.10       |
| eslint                      | 9.39.5       |
| node                        | 24.17.0      |
| npm                         | 11.18.0      |
| OS                          | macOS 26.5.2 |

(ESLint 10 is untested here — `eslint-plugin-import@2.32.0`'s peer range stops at `^9`.)

**Comment by ljharb** [comment]
This is a fundamental problem with the style of node_modules layout that "linked" (and yarn PNP, etc) provide.

I agree there's ways to fix it without a peer dep - but the intention here is to not need to explicitly declare it. In other words, the very thing this node_modules layout is trying to "fix" is the _feature_ that this plugin depends on. It's just not a compatible layout with the entirety of the npm ecosystem.
